File: 146.lru-cache.js
/*
 * @lc app=leetcode id=146 lang=javascript
 *
 * [146] LRU Cache
 */

// @lc code=start
/**
 * @param {number} capacity
 */

class Node {
  constructor(key, val, next = null, prev = null) {
    this.key = key
    this.val = val
    this.next = next
    this.prev = prev
  }
}

var LRUCache = function(capacity) {
  this._capacity = capacity
  this._cache =  new Map()
  this._head = new Node()
  this._tail = new Node()
  this._head.next = this._tail
  this._tail.prev = this._head
};

LRUCache.prototype.remove = function(node) {
  let prevNode = node.prev
  let nextNode = node.next
  prevNode.next = nextNode
  nextNode.prev = prevNode
}

LRUCache.prototype.insert = function(node) {
  let prevNode = this._head
  let nextNode = this._head.next
  prevNode.next = node
  nextNode.prev = node
  node.prev = prevNode
  node.next = nextNode
}

/** 
 * @param {number} key
 * @return {number}
 */
LRUCache.prototype.get = function(key) {
  if (this._cache.has(key)) {
    this.remove(this._cache.get(key))
    this.insert(this._cache.get(key))
    return this._cache.get(key).val
  }
  return -1
};

/** 
 * @param {number} key 
 * @param {number} value
 * @return {void}
 */
LRUCache.prototype.put = function(key, value) {
  if (this._cache.get(key)) {
    this.remove(this._cache.get(key))
  }
  this._cache.set(key, new Node(key, value))
  this.insert(this._cache.get(key))
  if (this._cache.size > this._capacity) {
    // remove from the list and delete LRU from cache
    let lru = this._tail.prev
    this.remove(lru)
    this._cache.delete(lru.key)
  }
};

/** 
 * Your LRUCache object will be instantiated and called as such:
 * var obj = new LRUCache(capacity)
 * var param_1 = obj.get(key)
 * obj.put(key,value)
 */
// @lc code=end

File: cpp/146.lru-cache.cpp
/*
 * @lc app=leetcode id=146 lang=cpp
 *
 * [146] LRU Cache
 */

// @lc code=start
class LRUCache
{
public:
    struct Node
    {
        int val;
        int key;
        Node *prev;
        Node *next;
    };
    struct DoubleLL
    {
        Node *head;
        Node *tail;
        DoubleLL()
        {
            head = new Node{};
            tail = new Node{};
            head->next = tail;
            tail->prev = head;
        }
        void add(Node *node)
        {
            Node *nextNode = head->next;
            head->next = node;
            node->prev = head;
            node->next = nextNode;
            nextNode->prev = node;
        }
        void remove(Node *node)
        {
            Node *prevNode = node->prev;
            Node *nextNode = node->next;
            prevNode->next = nextNode;
            nextNode->prev = prevNode;
        }
    };

    int maxCapacity;
    int currCapacity = 0;
    unordered_map<int, Node *> lookupMap;
    DoubleLL linkedList;

    LRUCache(int capacity)
    {
        // initialize doubly linked list
        // hashmap of pointers with key as key
        // and value as pointer to linked list
        // if capacity exceeded, remove tail
        // of linked list
        // on gets, shift linked list node
        // to head
        // O(k) get O(k) put
        maxCapacity = capacity;
    }

    int get(int key)
    {
        if (lookupMap[key])
        {
            // shift node to head position
            Node *currNode = lookupMap[key];
            linkedList.remove(currNode);
            linkedList.add(currNode);
            return lookupMap[key]->val;
        }
        else
        {
            return -1;
        }
    }

    void put(int key, int value)
    {
        // check capacity
        if (currCapacity == maxCapacity)
        {
            // check if we are replacing an existing node
            if (lookupMap[key])
            {
                linkedList.remove(lookupMap[key]);
                // add
                Node *newNode = new Node{value, key};
                linkedList.add(newNode);
                lookupMap[key] = newNode;
            }
            else
            {
                // remove last node
                Node *lastNode = linkedList.tail->prev;
                linkedList.remove(lastNode);
                lookupMap.erase(lastNode->key);
                // add
                Node *newNode = new Node{value, key};
                linkedList.add(newNode);
                lookupMap[key] = newNode;
            }
        }
        else
        {
            // remove prev value if key already set
            if (lookupMap[key])
            {
                linkedList.remove(lookupMap[key]);
                // add
                Node *newNode = new Node{value, key};
                linkedList.add(newNode);
                lookupMap[key] = newNode;
            }
            else
            {
                // add
                Node *newNode = new Node{value, key};
                linkedList.add(newNode);
                lookupMap[key] = newNode;
                currCapacity++;
            }
        }
    }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */
// @lc code=end
